src/security/context.cpp

```diff
@@ -643,16 +694,197 @@ class PolFinalWafCtx : public PolTaskCtx<PolFinalWafCtx> {
+    }
+
+    st = transition_to_stage(stage::COLLECTING_ON_REQ_DATA);
+  } else if (st == stage::COLLECTING_ON_REQ_DATA) {
+    // check if we have enough data to run the WAF
+    size_t new_size = filter_ctx_.out_total;
+    bool is_last = filter_ctx_.found_last;
+    for (auto *cl = in; cl; cl = cl->next) {
+      new_size += cl->buf->last - cl->buf->pos;
+      is_last = is_last || cl->buf->last_buf;
+    }
+
+    bool run_waf = new_size >= kMaxFilterData || is_last;
+    if (run_waf) {
+      // do not consume the buffer so that this filter is not called again
+      if (buffer_chain(*request.pool, in, false) != NGX_OK) {
+        return NGX_HTTP_INTERNAL_SERVER_ERROR;
+      }
+
+      if (filter_ctx_.out_total == 0) {
+        ngx_log_debug0(NGX_LOG_DEBUG_HTTP, request.connection->log, 0,
+                       "no data to run WAF on");
+        transition_to_stage(stage::AFTER_ON_REQ_WAF);
+        goto pass_downstream;
+      }
+
+      ngx_log_debug2(NGX_LOG_DEBUG_HTTP, request.connection->log, 0,
+                     "running WAF on %lu bytes of data (found last: %s)",
+                     new_size, is_last ? "true" : "false");
+
+      PolReqBodyWafCtx &task_ctx =
+          PolReqBodyWafCtx::create(request, *this, span);
+
+      transition_to_stage(stage::SUSPENDED_ON_REQ_WAF);
+
+      auto *conf = static_cast<datadog_loc_conf_t *>(
+          ngx_http_get_module_loc_conf(&request, ngx_http_datadog_module));
+
+      if (task_ctx.submit(conf->waf_pool)) {
+        ngx_log_debug(NGX_LOG_DEBUG_HTTP, request.connection->log, 0,
+                      "posted request body waf req post task");
+      } else {
+        transition_to_stage(stage::AFTER_ON_REQ_WAF);
+        ngx_log_error(NGX_LOG_NOTICE, request.connection->log, 0,
+                      "posted request body waf req post task failed. Passing "
+                      "data to downstream filters immediately");
+        goto pass_downstream;
+      }
+    } else {  // !run_waf; we need more data
+      if (buffer_chain(*request.pool, in, true) != NGX_OK) {
+        return NGX_HTTP_INTERNAL_SERVER_ERROR;
+      }
+    }
+  } else if (st == stage::AFTER_ON_REQ_WAF ||
+             st == stage::AFTER_ON_REQ_WAF_BLOCK) {
+    if (filter_ctx_.out) {  // first call after WAF ended
+      ngx_log_debug1(NGX_LOG_DEBUG_HTTP, request.connection->log, 0,
+                     "first filter call after WAF ended, req refcount=%d",
+                     request.main->count);
+      if (buffer_chain(*request.pool, in, false) != NGX_OK) {
+        return NGX_HTTP_INTERNAL_SERVER_ERROR;
+      }
+
+    pass_downstream:
+      // pass saved buffers downstream
+      auto rc = ngx_http_next_request_body_filter(&request, filter_ctx_.out);
+
+      for (auto *cl = filter_ctx_.out; cl;) {
+        auto *ln = cl;
+        cl = cl->next;
+        ngx_free_chain(request.pool, ln);
+      }
+      filter_ctx_.out = nullptr;
+      filter_ctx_.out_last = &filter_ctx_.out;
```

Comment:
`FilterCtx` has no destructor, so the data inside it wouldn't be freed by the destruction of the context; however, it the chain links are allocated as part of the request pool and would be freed then anyway. In general it's best to return data used to the request pool so it can be reused. I do notice though, that I don't free the temp buffers allocated in `buffer_chain` when `consume == true`; although not critical, that could be improved here.

Reply:
> FilterCtx has no destructor, so the data inside it wouldn't be freed by the destruction of the context

That's my point, this logic is better located in the `FilterCtx` destructor and the memory will be free on the non critical path. Or best, let the pool manage that.

Reply:
A couple of points:

* I was mistaken about the buffers; those can't be returned to the pool, only possibly reused by keeping freelists (e.g. with `ngx_chain_update_chains`). We're not interested in reusing buffers though, because we have an initial phase where we only allocate buffers and then after we pass them downstream they're not needed anymore. That is, we don't have allocations after the deallocations. So the memory occupied by those buffers can only be reclaimed when the pool is destroyed on request end.
* The chain links CAN be returned to the pool though, as they have a fixed size and `ngx_pool_t` keeps a freelist for them. If we released the chain links only at the end of the request, we'd be increasing memory consumption unnecessarily.